### ISSUE TITLE
Inheritance of models in APSIM development (e.g. Report)

### ISSUE THREAD

**Issue by lie112** [APSIMInitiative/ApsimX#6640]
I am checking on the policy and best practice for inheritance of models in APSIM next gen.

There are clear advantages to inheriting classes and polymorphism in that there is a single, reusable source of code logic and inherited classes can add to this functionality or provide their own implementation. A clean implementation of inheritance also avoids the problem of code changes not being propagated when duplicate code is added to the system to replicate functionality. There can also be disadvantages in keeping the software clean if not done properly.

I have added some inheritance within the CLEM model (individual RuminantFemale is inherited from Ruminant) and there is also need to build on the functionality of some of the APSIM components (e.g. Reports). There are some problems when trying to inherit some models.

Problems

- Models have not been implemented for inheritance with virtual or abstract classes and logic
- The binding of EventSubscribe does not consider base class methods
- You cannot ignore a property of the inherited model in the property presenter if the base class has description attribute (can check for empty description as well as description present to allow inherited description text to determine outcome). Can also be solved by using "new" if the property is not needed, but this is sloppy.

Allow inheritance of Report.cs
CLEM provides users with reports that automatically generate the variables and events list and require some additional properties and presenters. This has been achieved by inheriting the Report model and providing a new "Commencing" method. Some additional changes are needed to fix the need to duplicate a range of methods in the inherited class (e.g. DoOutput, static SortColumnsOfDataTable, FindVariableMembers) which work with the base private variables that are not accessible by the derived class.

**Comment by hol430** [comment]
The [coding style guidelines](https://apsimnextgeneration.netlify.app/development/software/codingstyle/) specify that it's "discouraged". I don't want to start a flame war but my 2c is that I think it can be useful in some cases but is very easy to abuse. I don't think a blanket ban is desirable - in fact, I'm using inheritance at the moment in my autodocs refactor. On the other hand, there are other mechanisms to facilitate code reuse/sharing (such as composition) which I think would be preferable in a lot of cases.

Regarding your problems:1 and 2 are easily rectified (2 is arguably a bug rather than a design decision). 3 could be fixed by setting Inherited to false on the AttributeUsage attribute on the DescriptionAttribute class (I think). Which seems like a reasonable solution to me.

**Comment by jbrider** [comment]
@lie112 Coming from C++ I was far more comfortable with inheritance than composition but have found it's not too difficult to make the shift.  If you have the need to copy+paste functions then look to extract those functions into a different standalone object that can be added in.  
I have dealt with the intricacies of EventSubscribe within inheritance and it really does not lend itself to more readable code.

**Comment by hol353** [comment]
I'm not against using inheritance. I'm just trying to avoid the APSIM 7.10 problem in Plant (and elsewhere) where there is a long chain of inheritance and even multiple inheritance I think. This then leads to functionality scattered across many files and you can't find what you're looking for easily. This is also a criticism of PMF where there any **lots** of child functions and a plant model is scattered all over the place. I also agree with @jbrider that events hinder readable code. In some ways I like the old FORTRAN days were a crop model is all in one file - simpler. 

**What we're trying to achieve is 'readable and maintainable code'. We haven't found the perfect design philosophy yet - it probably doesn't exist.**

@lie112 set up a meeting with @hol430 and I and we'll have a look at what you're trying to do.

**Comment by lie112** [comment]
Thanks for all the feedback. I wasn't aware of the "composition" coding style and will consider it in future. In fact I have already used that approach without realising. I do realise the benefits of inheritance in the code behind CLEM and probably use types to identify classes where Interfaces would be a better approach that I am now regularly using Interfaces. This seems far more relevant to all the functionality that controls the operation of the CLEM model and interaction with the APSIM UI than the core model code where there is limited inheritance.

It is really useful to learn of the issues of the past and how they can be avoided having not experienced APSIM classic. Coming in new to the APSIM next gen code I found all the interfaces equally difficult to get my head around (especially the model view style of the UI). It wasn't the style, it was just getting your head around all the names etc, but in general I have found APSIM clean and good to use and expand functionality an options without breaking anything for others. 

The event based logic is great for organising the order of things happening and an improvement on straight through code of fortran and old vb days, allowing individual components to respond to events and allowing priority to be customised in CLEM by the order components (models) are supplied in the tree. The resource - activity - utilities approach of CLEM does keep code together in logical places, but I agree the advanced user does need to understand the nested inheritance providing the functionality of various components. 

I have applied all the suggestions to allow Reports to be inherited making for much cleaner code in my custom reports as now all I need to supply are any properties and a oncommencing method to set up the VariableNames and EventNames arrays. This is now much cleaner and I enjoy learning new simple approaches that quickly allow APSIM to handle the new ideas such as the setting of Inherited to false on the AttributeUsage. I also need to upgrade my activities performed report to use the new grid table rather than the old treeview and I'd like to implement font-icons rather than raster images for zoomability. Again this just needs to provide data to the table and add a bit of logic for displaying something in each cell rather than the value.

I can pull all changes making reports inheritable and the simplified CLEM reports for perusal and acceptance once my current pull request is accepted as it's in the same branch. I seriously need to update my skills in GIT and separating concern in how to provide changes to the bigger solution when working on bugs and improvements while developing new sections.

I noticed there is a previous classDiagram.cs in the solution. Do we provide any class diagrams for users to understand all the relationships between classes or is this handled in the auto doc? I would like to generate a CLEM class diagram for display in the advanced users section of the CLEM help if anyone has ideas on the best tools for that.
